### PR TITLE
#167377557 Implement bookmark article functionality

### PR DESCRIPTION
**Pull request by kennyedward**
#### What does this PR do?
This pull request enables a user to add or remove an article from a bookmark

#### Description of Task to be completed
- add an article to bookmark
- remove an article from a bookmark

#### How should this be manually tested?
 - git clone this repository
 - ```cd``` into the repo & run ```npm install```
- Update the **.env** with the **.env.sample** file on this repo
- Using mocha, run ```npm run test```

#### Any Background Context You want to provide?
N/A

#### What are the relevant pivotal tracker stories?

[#167377557](https://pivotaltracker.com/story/show/167377557)

#### Screenshots (if appropriate)
Bookmark Added Successfully
Bookmark Already Exists
Article Removed From Bookmark
Article Not Found

#### Questions:
N/A

### CODE REVIEW

File: src/controllers/article.js [hunk 1]
```diff
@@ -21,5 +22,29 @@ export default {
...
+      *
+      * @return {Object} - callback that execute the createBookmark controller
+      */
+  createBookmark: async (request) => {
+    const { articleId } = request.params;
+    const userId = request.body.user || {};
+
+    await Articles.getExistingArticle(articleId);
+
+    const existingBookmark = await Bookmarks.getExistingBookmark(articleId);
```
Comment (Merkll) [inline src/controllers/article.js]: Good job Kenny, but this code doesn't validate if the bookmark belongs to the user requesting for the bookmark. Also, you don't need to verify the user as the auth middleware does that

**Review by Merkll — COMMENTED** [review]
Check the review I left

File: src/controllers/article.js (hunk 1 above)
Comment (kennyedward) [reply src/controllers/article.js]: Thanks for pointing that out. I have refactored the code to reflect these changes.

File: .env.example [hunk 2]
```diff
@@ -1,15 +1,15 @@
+PORT=
```
Comment (iverenshaguy) [inline .env.example]: Please update this .env and put back every variable that was removed

File: src/controllers/article.js [hunk 3]
```diff
@@ -22,5 +23,53 @@ export default {
...
+
+    const articleObject = await Articles.getExistingArticle(articleId);
+
+    const existingBookmark = await Bookmarks.getExistingBookmark(articleId, userId);
+
+    if (existingBookmark) throw new ApplicationError(409, 'bookmark already added');
+
+    const newBookmark = await articleObject.createBookmark({ userId });
+
+    return response.status(201).json({ data: newBookmark, message: 'bookmark added succesfully' });
```
Comment (iverenshaguy) [inline src/controllers/article.js]: Please let this error start with a Capital letter. I noticed a lot of people using Sentence case and decided that it was better to use what most apps use.
```suggestion
    return response.status(201).json({ data: newBookmark, message: 'Bookmark added succesfully' });
```

File: src/controllers/article.js [hunk 4]
```diff
@@ -22,5 +23,53 @@ export default {
...
+    const { articleId } = request.params;
+    const { user: userId } = request.body;
+
+    const existingBookmark = await Bookmarks.getExistingBookmark(articleId, userId);
+
+    if (!existingBookmark) throw new NotFoundError();
+
+    await Bookmarks.destroy({ where: { articleId, userId } });
+
+    return response.status(200).json({ message: 'article removed from bookmark' });
```
Comment (iverenshaguy) [inline src/controllers/article.js]: Same feedback as above

File: src/models/Articles.js [hunk 5]
```diff
@@ -114,4 +114,22 @@ export default class Articles extends Model {
...
+   * Get existing article
+   *
+   * @static
+   * @memberof Articles
+   *
+   * @param {string} articleId
+   *
+   * @return {Object | void} - details of existing article
+   */
+  static async getExistingArticle(articleId) {
```
Comment (iverenshaguy) [inline src/models/Articles.js:152]: Good one

File: src/models/Bookmarks.js [hunk 6]
```diff
@@ -55,4 +57,21 @@ export default class Bookmarks extends Model {
...
+   * @static
+   * @memberof Bookmarks
+   *
+   * @param {string} articleId - id of the article to bookmark
+   * @param {string} userId - id of user
+   *
+   * @return {Object | void} - details of existing bookmark
+   */
+  static async getExistingBookmark(articleId, userId) {
+    const bookmark = await Bookmarks.findOne({ where: { articleId, userId } });
```
Comment (iverenshaguy) [inline src/models/Bookmarks.js:73]: Good job

File: src/models/Users.js [hunk 7]
```diff
@@ -113,4 +113,32 @@ export default class Users extends Model {
...
+   *  Model associations
+   *
+   * @static
+   * @memberof Users
+   *
+   * @param {any} models all models
+   *
+   * @returns {void} no return
+   */
+  static associate(models) {
```
Comment (iverenshaguy) [inline src/models/Users.js]: Thanks for updating

**Review by iverenshaguy — APPROVED** [review]
Great work. Almost no review. I'm proud of this work. Please update your work according to my review. Thank you

File: src/controllers/article.js [hunk 8]
```diff
@@ -22,5 +23,53 @@ export default {
     const articleData = await Articles.createComment({ articleId, comment });
 
     return response.status(201).json({ status: 'success', data: articleData, message: 'Comment added succesfully' });
+  },
+
+  /**
+      * controller for adding article to bookmark
```
Comment (iverenshaguy) [inline src/controllers/article.js]: Please align this properly

**Review by Nennyfills — DISMISSED** [review]

File: src/controllers/article.js (hunk 8 above)
Comment (kennyedward) [reply src/controllers/article.js]: Thanks for the review. I have implemented all the feedback.

**Comment by iverenshaguy** [comment]
For some reason `status` isn't part of your response body. Please add it. Thank you

**Pull request merged**